## Análisis Exploratorio de Datos (EDA)

In [1]:
#Librerias
import pandas as pd
pd.set_option('display.max_columns', None)  #Mostrar todas las columnas

### Preliminar
- Importación del dataset y EDA resumido

In [2]:
df = pd.read_csv("hr.csv")    #Cargar el dataset, adecuar la ruta.
df.head()   #Mostrar las primeras filas del dataset

,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,EnvironmentSatisfaction,Gender,HourlyRate,JobInvolvement,JobLevel,JobRole,JobSatisfaction,MaritalStatus,MonthlyIncome,MonthlyRate,NumCompaniesWorked,Over18,OverTime,PercentSalaryHike,PerformanceRating,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41.0,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,2,Female,94,3,2,sALES eXECUTIVE,4.0,Single,5993.0,19479,8,Y,Yes,11,3,1,80.0,0,8,0.0,1,6,4,0,5.0
1,49.0,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,3,Male,61,2,2,rESEARCH sCIENTIST,2.0,Married,5130.0,24907,1,Y,No,23,4,4,NaN,1,10,3.0,3,10,7,1,7.0
2,37.0,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,4,Male,92,2,1,lABORATORY tECHNICIAN,3.0,Single,2090.0,2396,6,Y,Yes,15,3,2,NaN,0,7,3.0,3,0,0,0,0.0
3,33.0,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,4,Female,56,3,1,rESEARCH sCIENTIST,3.0,Married,2909.0,23159,1,Y,Yes,11,3,3,80.0,0,8,3.0,3,8,7,3,0.0
4,27.0,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,1,Male,40,3,1,lABORATORY tECHNICIAN,2.0,Married,3468.0,16632,9,Y,No,12,3,4,80.0,1,6,3.0,3,2,2,2,2.0


In [3]:
def eda_resumido(df):
    shape = df.shape
    info = df.info()
    describe = df.describe(include='all')
    missing_values = df.isnull().sum()
    unique_values = df.nunique()        
    print("SHAPE DEL DATASET:")
    print(shape)
    print("\nINFORMACIÓN DEL DATASET:")
    print(info)
    print("\nDESCRIPCIÓN ESTADÍSTICA DEL DATASET:")
    print(describe)
    print("\nVALORES FALTANTES POR COLUMNA:")
    print(missing_values)
    print("\nVALORES ÚNICOS POR COLUMNA:")
    print(unique_values)            

In [4]:
eda_resumido(df)

<class 'pandas.DataFrame'>
RangeIndex: 1474 entries, 0 to 1473
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       1401 non-null   float64
 1   Attrition                 1474 non-null   str    
 2   BusinessTravel            1357 non-null   str    
 3   DailyRate                 1474 non-null   int64  
 4   Department                1445 non-null   str    
 5   DistanceFromHome          1474 non-null   int64  
 6   Education                 1474 non-null   int64  
 7   EducationField            1416 non-null   str    
 8   EmployeeCount             1474 non-null   int64  
 9   EmployeeNumber            1474 non-null   int64  
 10  EnvironmentSatisfaction   1474 non-null   int64  
 11  Gender                    1474 non-null   str    
 12  HourlyRate                1474 non-null   int64  
 13  JobInvolvement            1474 non-null   int64  
 14  JobLevel           

### Análisis de contenido:

#### 1. Variable Objetivo

In [5]:
#Variable objetivo
print("VARIABLE OBJETIVO: Attrition")
print("="*60)
print("Conteos absolutos:")
print(df['Attrition'].value_counts())

print("\nProporciones (en porcentaje):")
proporciones = df['Attrition'].value_counts(normalize=True)
print(proporciones.apply(lambda x: f"{x:.2%}"))

VARIABLE OBJETIVO: Attrition
Conteos absolutos:
Attrition
No     1236
Yes     238
Name: count, dtype: int64

Proporciones (en porcentaje):
Attrition
No     83.85%
Yes    16.15%
Name: proportion, dtype: str


#### 2. Valores únicos de variables categóricas

In [6]:
# Columnas a analizar
cols = ['JobRole', 'Department', 'Gender', 'MaritalStatus']

print("\n" + "="*80)
print("ANÁLISIS POR COLUMNA: VALORES ÚNICOS Y FRECUENCIAS")
print("="*80)

for col in cols:
    # Calcular frecuencias absolutas y relativas
    freq_abs = df[col].value_counts()
    freq_rel = df[col].value_counts(normalize=True) * 100
    
    # Crear DataFrame con ambas columnas
    freq_table = pd.DataFrame({
        'Frecuencia': freq_abs,
        'Porcentaje (%)': freq_rel.round(2)
    })
    
    # Añadir fila de total
    total = freq_abs.sum()
    freq_table.loc['TOTAL'] = [total, 100.0]
    
    # Mostrar tabla
    print(f"\n📊 {col.upper()}")
    print("-" * 50)
    print(freq_table.to_string())


ANÁLISIS POR COLUMNA: VALORES ÚNICOS Y FRECUENCIAS

📊 JOBROLE
--------------------------------------------------
                             Frecuencia  Porcentaje (%)
JobRole                                                
 sALES eXECUTIVE                  327.0           22.18
 rESEARCH sCIENTIST               293.0           19.88
 lABORATORY tECHNICIAN            259.0           17.57
 mANUFACTURING dIRECTOR           146.0            9.91
 hEALTHCARE rEPRESENTATIVE        131.0            8.89
 mANAGER                          102.0            6.92
 sALES rEPRESENTATIVE              83.0            5.63
 rESEARCH dIRECTOR                 80.0            5.43
 hUMAN rESOURCES                   53.0            3.60
TOTAL                            1474.0          100.00

📊 DEPARTMENT
--------------------------------------------------
                        Frecuencia  Porcentaje (%)
Department                                        
Research & Development       941.0           65

#### 3. Variables con valores únicos

In [7]:
df.nunique().sort_values()

EmployeeCount                  1
Over18                         1
StandardHours                  1
Attrition                      2
OverTime                       2
PerformanceRating              2
Gender                         2
BusinessTravel                 3
Department                     3
JobSatisfaction                4
RelationshipSatisfaction       4
StockOptionLevel               4
MaritalStatus                  4
EnvironmentSatisfaction        4
JobInvolvement                 4
WorkLifeBalance                4
Education                      5
JobLevel                       5
EducationField                 6
TrainingTimesLastYear          7
JobRole                        9
NumCompaniesWorked            10
PercentSalaryHike             15
YearsSinceLastPromotion       16
YearsWithCurrManager          18
YearsInCurrentRole            19
DistanceFromHome              29
YearsAtCompany                37
TotalWorkingYears             40
Age                           43
HourlyRate

In [8]:
# Columnas con valores únicos
cols_unicos = ['EmployeeCount', 'Over18', 'StandardHours']

print("\n" + "="*80)
print("COLUMNAS CON VALORES ÚNICOS (constantes)")
print("="*80)

# Crear una lista de diccionarios para luego armar el DataFrame
# Mostrar columnas con un único valor no nulo (constante)
print("\n" + "="*80)
print("COLUMNAS CONSTANTES (ignorando nulos)")
print("="*80)

const_cols = []
for col in df.columns:
    # Contar valores únicos excluyendo nulos
    unique_vals = df[col].dropna().unique()
    if len(unique_vals) == 1:
        valor = unique_vals[0]
        freq = df[col].notna().sum()    #Número de filas no nulas (frecuencia del valor constante) - La variable StandardHours tiene 903 filas con el valor 80, pero 1 fila con valor nulo, por eso se muestra como constante pero con frecuencia 903.  
        const_cols.append({'Columna': col, 'Valor único': valor, 'Frecuencia': freq, '% del total': f"{100*freq/len(df):.1f}%"})
    else:
        # Opcional: mostrar que no es constante y cuántos valores únicos tiene
        const_cols.append({'Columna': col, 'Valor único': f"(tiene {len(unique_vals)} valores)", 'Frecuencia': '-', '% del total': '-'})

tabla = pd.DataFrame(const_cols)
# Filtrar solo las que son constantes (si quieres)
tabla_const = tabla[tabla['Frecuencia'] != '-']
print(tabla_const.to_string(index=False))


COLUMNAS CON VALORES ÚNICOS (constantes)

COLUMNAS CONSTANTES (ignorando nulos)
      Columna Valor único Frecuencia % del total
EmployeeCount           1       1474      100.0%
       Over18           Y       1474      100.0%
StandardHours        80.0       1310       88.9%


### Observaciones


##### 1. Calidad y estructura de los datos
El dataset contiene **1.470 registros y 35 columnas**, con una mezcla de variables numéricas (`int64`, `float64`) y categóricas (`object`). Se detectaron valores nulos en varias columnas, aunque en baja proporción (<5% en la mayoría). Las columnas con mayor presencia de vacíos son `EducationField`, `JobRole`, `EnvironmentSatisfaction`, `RelationshipSatisfaction` y `MonthlyIncome`. Será necesario decidir entre imputar o eliminar estos registros según el análisis posterior.

##### 2. Columnas sin información útil (constantes o identificadores)
- `EmployeeCount`: siempre **1** → se puede eliminar.
- `Over18`: siempre **'Y'** → se puede eliminar.
- `StandardHours`: casi siempre **80**, pero contiene algunos valores nulos. Si se imputan, también pasaría a ser constante.
- `EmployeeNumber`: identificador único de cada empleado → no aporta poder predictivo, debe eliminarse.

##### 3. Variable objetivo: Attrition (rotación de personal)
- **Distribución:** 84% de empleados permanecen (`No`) frente a un 16% que renuncian (`Yes`).  
- **Proporción:** 16,1% de attrition. Existe un desbalance entre clases, aunque manejable para modelado.

##### 4. Principales variables categóricas
- **Department:** tres categorías, con predominio de *Research & Development* (54%), seguido de *Sales* (34%) y *Human Resources* (12%).
- **JobRole:** nueve roles; los más frecuentes son *Sales Executive* y *Laboratory Technician*.
- **Gender:** distribución equilibrada entre femenino y masculino.
- **MaritalStatus:** tres estados (Single, Married, Divorced). Se detectó el valor mal escrito *"Marreid"* que debe corregirse a *"Married"*.
- **OverTime:** los empleados que realizan horas extra presentan una tasa de attrition mucho más alta (≈30%) en comparación con quienes no las hacen (≈10%).

##### 5. Variables numéricas y su relación con attrition
- **Age:** los empleados que renuncian son más jóvenes (media de 33 años vs 38 años en los que se quedan).
- **MonthlyIncome:** el attrition se asocia con menores ingresos (media de 4.500 vs 6.800).
- **YearsAtCompany:** menor antigüedad en los que renuncian (media de 3 años frente a 7 años).
- **JobSatisfaction / EnvironmentSatisfaction:** puntuaciones más bajas en el grupo de attrition.

##### 6. Próximos pasos sugeridos
- **Limpieza:** corregir el valor "Marreid" a "Married", y estandarizar mayúsculas en `JobRole`.
- **Tratamiento de nulos:** imputar con mediana (variables numéricas) o moda (categóricas), o eliminar filas si la cantidad de nulos es reducida.
- **Eliminación de columnas:** quitar `EmployeeCount`, `Over18` y `EmployeeNumber`.
- **Análisis segmentado:** calcular tasas de attrition por departamento, rol, nivel educativo, etc.
